R/aggregate.R

```diff
@@ -180,43 +143,380 @@ agg_climate_data <- function(year, data_raster, climate_var, daily_agg, trans =
+  list_length <- length(poly_orders) # How many lists are in the final object
+  list_names <- sapply(1:list_length, FUN=function(x){paste("order", poly_orders[x], sep="_")})
 
-  no_cores <- parallel::detectCores() - 1 # Calculate the number of cores. Leave one in case something else needs to be done on the same computer at the same time.
-  cl <- parallel::makeCluster(no_cores, type="FORK") # Initiate cluster. "FORK" means bring everything in your current environment with you.
-  sum_by_poly_multiyear <- parallel::parLapply(cl, years, agg_climate_data, data_raster, climate_var, daily_agg, trans, trans_specs, geoweights_table)
-  parallel::stopCluster(cl)
+  # For each daily layer, raise the value to degree, degree-1, degree-2 etc. until 1
+  r <- lapply(poly_orders, FUN=function(x){clim_daily ^ x})
 
-  sum_by_poly_multiyear <- data.table::rbindlist(sum_by_poly_multiyear)
 
+    ## Function: Set names of data table by month, change from wide to long format, rename based on polynomial orders
+  create_dt <- function(x){
+
+    # Should output raster cells x/y with 365 days as column names
+    dt <- as.data.table.raster(r[[x]], xy=TRUE)
+
+    # Set column names with months
+    new_names <- c('x', 'y', layer_names)
+    data.table::setnames(dt, new_names)
 
-  return(sum_by_poly_multiyear)
+    # Change from wide to long format
+    dt = data.table::melt(dt, id.vars = c("x", "y"))
+
+    # Update variable names
+    var_names <- c('date', list_names[x])
+    data.table::setnames(dt, old=c('variable', 'value'), new=var_names)
+  }
+
+  # Make each raster layer a data.table
+  list_dt <- lapply(1:list_length, create_dt)
+
+  # Merge all data tables together
+  clim_dt <- list_dt[[1]]
+  for(i in 2:list_length){
+    dt_m <- list_dt[[i]]
+    clim_dt <- merge(clim_dt, dt_m, by=c('x', 'y', 'date'))
+  }
+
+  # Aggregate by polygon
+  sum_by_poly <- after_trans(clim_dt, weights_dt, list_names, second_weights, agg_to)
+
+  return(sum_by_poly)
 
-  # Done
-  message(crayon::green('Done'))
 }
+
+#' @export
+staggregate_spline <- function(data_raster, climate_var, daily_agg, knot_locs, geoweights_table, second_weights, agg_to = "month"){
+
+  # Get climate data as a data table and aggregate to daily values
+  setup_list <- before_trans(data_raster, climate_var, daily_agg, geoweights_table, second_weights)
+
+  clim_daily <- setup_list[[1]]
+  weights_dt <- setup_list[[2]]
+  layer_names <- setup_list[[3]]
+
+
+  # Spline transformation
+  knot_locs <- sort(knot_locs)
+  num_knots <- length(knot_locs)
+  list_length <- num_knots - 2
+  list_names <- sapply(0:list_length, FUN=function(x){if(x == 0){paste(climate_var)}else{paste("term", x, sep="_")}})
+
+
+  # Define spline function
+  get_spline <- function(x){
+
+    # Make first raster returned just the climate variable to preserve it's column in the resulting data table
+    if(x == 0){
+      return(clim_daily)
+    }
+    # Add in spline terms, all of which are 0 if negative
```

Comment:
@tliddell4 flagging that i want to review this chunk of code carefully (the splines) but don't have time right this second. will try later tonight.